### ISSUE TITLE
spring-boot-maven-plugin:build-image fails on Docker Desktop / Windows due missing authentication

### ISSUE THREAD

**Issue by eidottermihi** [spring-projects/spring-boot#36671]
When running the `build-image` goal of spring-boot-maven-plugin on Windows with Docker Desktop, the build fails when the plugin tries to pull the buildpacks builder image via Docker API.

```
[INFO] --- spring-boot-maven-plugin:3.1.2:build-image-no-fork (build-image) @ appswitcher-server ---
[INFO] Building image 'docker.io/itatm/appswitcher-server:1.0.3-SNAPSHOT'
[INFO]
[INFO]  > Pulling builder image 'docker.io/paketobuildpacks/builder:base' 100%
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.199 s
[INFO] Finished at: 2023-08-02T10:51:59+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin:3.1.2:build-image-no-fork (build-image) on project appswitcher-server: 
  Execution build-image of goal org.springframework.boot:spring-boot-maven-plugin:3.1.2:build-image-no-fork failed: 
    Docker API call to 'localhost/v1.24/images/create?fromImage=docker.io%2Fpaketobuildpacks%2Fbuilder%3Abase' failed with status code 407 "Proxy Authentication Required" -> [Help 1]
```
(full maven debug log see attachment)

This seems to be related to an authentication issue, because the `docker.log` (located `%LOCALAPPDATA%\Docker\log\vm`) contains this error:

```
[2023-08-02T08:51:59.764710000Z][docker][I] [2023-08-02T08:51:59.764357000Z][lifecycle-server][I] missing X-Registry-Auth header
[2023-08-02T08:51:59.764852000Z][docker][I] [2023-08-02T08:51:59.764593100Z][lifecycle-server][I] must be logged in to call /v1.24/images/create?fromImage=docker.io%2Fpaketobuildpacks%2Fbuilder%3Abase
```

To debug this issue further I manually issued the problematic Docker API call with cURL:

```bash
$ curl -X POST -v http://localhost:2375/v1.24/images/create?fromImage=docker.io%2Fpaketobuildpacks%2Fbuilder%3Atiny
* Connected to localhost (::1) port 2375 (#0)
> POST /v1.24/images/create?fromImage=docker.io%2Fpaketobuildpacks%2Fbuilder%3Atiny HTTP/1.1
> Host: localhost:2375
> User-Agent: curl/7.73.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 407 Proxy Authentication Required
< Content-Type: text/plain; charset=utf-8
< Date: Wed, 02 Aug 2023 08:11:10 GMT
< X-Content-Type-Options: nosniff
< Transfer-Encoding: chunked
<
{ [32 bytes data]
100    26    0    26    0     0   1238      0 --:--:-- --:--:-- --:--:--  1300Please run 'docker login'

* Connection #0 to host localhost left intact
```

As the response body of the problematic call contains `Please run 'docker login'` I also checked that my Docker `config.json` contains the relevant data:

```json
{
	"auths": {
		"ghcr.io": {},
		"https://index.docker.io/v1/": {}
	},
	"credsStore": "wincred"
}
```

So somehow this is also related or blocked by #25898  

Pulling the image via Docker CLI works without a problem:

```
$ docker pull docker.io/paketobuildpacks/builder:base
base: Pulling from paketobuildpacks/builder
Digest: sha256:17ea21162ba8c7717d3ead3ee3836a368aced7f02f2e59658e52029bd6d149e7
Status: Downloaded newer image for paketobuildpacks/builder:base
docker.io/paketobuildpacks/builder:base
```

My environment:

- Windows 10 22H2 (Build 19045.3208)
- Docker Desktop v4.19.0
- Spring Boot 3.1.2

[full_log.txt](https://github.com/spring-projects/spring-boot/files/12238382/full_log.txt)

**Comment by scottfrederick** [comment]
The Spring Boot plugin and the `curl` command both fail with `407 Proxy Authentication Required`. It appears there is a network proxy between the Docker daemon and the `docker.io` host that requires authentication. 

> So somehow this is also related or blocked by https://github.com/spring-projects/spring-boot/issues/25898

The linked issue is related to Docker repositories that require authentication in order to pull images. That's not the case here, as the [Paketo builder repository](https://hub.docker.com/r/paketobuildpacks/builder) is a [sponsored OSS repo](https://www.docker.com/community/open-source/application/) that does not require authentication and is not subject to rate limiting. It's a proxy that is requiring authentication, not the repository. 

> Pulling the image via Docker CLI works without a problem:

Have some proxy configuration been applied to the Docker CLI? 

There are no options on the Spring Boot build plugins to configure a proxy. There are [configuration options](https://docs.spring.io/spring-boot/docs/current/maven-plugin/reference/htmlsingle/#build-image.docker-daemon) that tell the plugins how to talk to the Docker daemon process, but that is different from the Docker daemon talking to a repository. 

Can you try some of the options documented by Docker [here](https://docs.docker.com/network/proxy/) and [here](https://docs.docker.com/desktop/settings/windows/#proxies) to configure the daemon for proxy?

**Comment by eidottermihi** [comment]
@scottfrederick thank you for looking into this issue!

> It appears there is a network proxy between the Docker daemon and the docker.io host that requires authentication.

That is correct, but our corporate proxy does _not_ require authentication. I've configured the proxy for the daemon [according to the docker docs](https://docs.docker.com/engine/reference/commandline/dockerd/#proxy-configuration) in `daemon.json`:

```json
{
  "proxies": {
    "http-proxy": "http://proxy.mycorp.com:80",
    "https-proxy": "http://proxy.mycorp.com:80",
    "no-proxy": "*.mycorp.com,mycorp.com"
  }
}
```

I still think it is related to missing authentification for docker.io:

If I'm using the Docker CLI and do a "docker logout" and then performing a docker pull, this then also fails:

```
$ docker logout
Removing login credentials for https://index.docker.io/v1/

$ docker pull docker.io/paketobuildpacks/builder:tiny
Error response from daemon: Please run 'docker login'
```

The error message for docker pull looks exactly the same as the error, when the maven plugin tries to pull this image:

```
[2023-08-03T06:19:35.610758000Z][docker][I] [2023-08-03T06:19:35.610417500Z][lifecycle-server][I] must be logged in to call /v1.42/images/create?fromImage=paketobuildpacks%2Fbuilder&tag=tiny
```

After performing the `docker login` for docker.io again, `docker pull` is working again.

If I'm configuring authentication for the builderRegistry directly in plugin configuration, it starts to work!

```xml
<plugin>
  <artifactId>spring-boot-maven-plugin</artifactId>
  <configuration>
    <docker>
      <builderRegistry>
        <username>my_dockerhub_user</username>
        <password>my_dockerhub_pat</password>
        <url>https://index.docker.io/v1/</url>
        <email>my_dockerhub_email</email>
      </builderRegistry>
    </docker>
<!-- ... --->
```

So It seems to me that docker.io requires auth - also for sponsored OSS images? Not sure if this is a maven plugin related issue then, but maybe this can be added to the docs?

Or https://github.com/spring-projects/spring-boot/issues/25898 comes back into play, as this would get the authentication from my local `config.json`?

**Comment by scottfrederick** [comment]
> So It seems to me that docker.io requires auth - also for sponsored OSS images? 

This is not the case. I have Docker Engine and Docker Desktop on Linux, and I can pull the Paketo builder images while logged out with both of those Docker daemons. Spring Boot has CI infrastructure that pulls Paketo images without authentication also. 

The flow here is: Docker client (CLI or Boot plugins) using the [Docker Engine API](https://docs.docker.com/engine/api/) -> Docker deamon (Docker Desktop in your case) at `localhost/v1.42` -> `docker.io` 

I believe the message `must be logged in to call /v1.42/images/create?fromImage=paketobuildpacks%2Fbuilder&tag=tiny` is the Docker daemon telling client using the API at `localhost/v1.42` (the Boot plugin in this case) that the client must be logged in or provide authentication before using the Docker Engine API. I do not think this is the `docker.io` host telling the daemon that authentication to the repository is required. The same thing goes for the `Error response from daemon: Please run 'docker login'` message from the Docker CLI. 

You could test this theory by using the Docker CLI or the Boot plugins to pull an image from a different repository to see if that changes anything. 

I don't know what would cause Docker Desktop on Windows to require any call to the Docker Engine API to be authenticated, but that appears to be what's going on.

**Comment by eidottermihi** [comment]
> You could test this theory by using the Docker CLI or the Boot plugins to pull an image from a different repository to see if that changes anything.

Apparently Docker Desktop on Windows requires authentication when pulling from `docker.io` - all other public registries work fine without authentication:

with docker.io: "docker login" required

```$ curl -X POST -v http://localhost:2375/v1.24/images/create?fromImage=docker.io%2Fpaketobuildpacks%2Fbuilder%3Atiny
* Connected to localhost (::1) port 2375 (#0)
> POST /v1.24/images/create?fromImage=docker.io%2Fpaketobuildpacks%2Fbuilder%3Atiny HTTP/1.1
< HTTP/1.1 407 Proxy Authentication Required
100    26    0    26    0     0   1529      0 --:--:-- --:--:-- --:--:--  1625Please run 'docker login'
```

with quay.io registry: works fine

```
curl -X POST -v http://localhost:2375/v1.24/images/create?fromImage=quay.io%2Fopenshift%2Forigin-cli
* Connected to localhost (::1) port 2375 (#0)
> POST /v1.24/images/create?fromImage=quay.io%2Fopenshift%2Forigin-cli HTTP/1.1
< HTTP/1.1 200 OK
<
100  2244    0  2244    0     0    269      0 --:--:--  0:00:08 --:--:--   548{"status":"Pulling from openshift/origin-cli","id":"v3.10.0"}
{"status":"Pulling fs layer","progressDetail":{},"id":"7dc0dca2b151"}
[....]
```

This is also the case if I'm using the Docker CLI:

```
$ docker pull docker.io/paketobuildpacks/builder:tiny
Error response from daemon: Please run 'docker login'

$ docker pull quay.io/openshift/origin-cli
Using default tag: latest
latest: Pulling from openshift/origin-cli
[...]
```

I will open an case at Docker Desktop support and hope they can shine some light into why the Docker Engine API on Docker Desktop for Windows requires authentication to pull public/sponsored OSS image from docker.io.

In the meantime I will use the "workaround" and supply `docker.io` builderRegistry credentials to the spring-boot-maven-plugin (see https://github.com/spring-projects/spring-boot/issues/36671#issuecomment-1663370400 ).

@scottfrederick Thanks for sticking with me! Closing the issue as it not related to Spring Boot and the Maven plugin itself.

**Comment by daggerok** [comment]
Faced with same issue on may mac
I authorized with corporate credentials and I don't have any proxy configuration
Only described workaround works when to spring-boot maven plugin we need to pass credentials
```
> mvn -v
Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937)
Maven home: /Users/maksim.kostromin/.homebrew/Cellar/maven/3.9.9/libexec
Java version: 23.0.1, vendor: Oracle Corporation, runtime: /Library/Java/JavaVirtualMachines/graalvm-23.jdk/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "15.2", arch: "aarch64", family: "mac"

> java --version
java 23.0.1 2024-10-15
Java(TM) SE Runtime Environment Oracle GraalVM 23.0.1+11.1 (build 23.0.1+11-jvmci-b01)
Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 23.0.1+11.1 (build 23.0.1+11-jvmci-b01, mixed mode, sharing)

Spring Boot 3.4.1
```

**Comment by fightin69th** [comment]
This is the first time I've responded but the following is what I have found to fix it.  As I had the same 407 issue. The issue stems from using the maven-spring-boot-plugin

Adding the following lines in the pom as part of the configuration section should fix it if you are using the unsecured daemon:

					<docker>
						<host>tcp://localhost:2375</host>
						<tlsVerify>false</tlsVerify>
					</docker>

*Host should be whatever settings you are using for your daemon. 
I hope this helps you as I spent a numerous amount of hours to finally get to it building with the paketo images again.

**Comment by Gyllsdorff** [comment]
_Just in case someone else find this issue later._ 

If you are using a company controlled Docker Desktop on your computer then it is possible that they have enabled something called ["enforced sign-in"](https://docs.docker.com/security/for-admins/enforce-sign-in/). That means that all requests going through your local device must be authenticated, if most cases that means that you have to do `docker login` once which stores your short lived credentials in you OS keychain but there are a few exceptions.

`docker login` (mostly) handles authentication for `docker` commands from the terminals, interactions through the local `/var/run/docker.sock` on the other hand must provide their own authentication credentials as they are not allowed to use the credentials that `docker login` created. That means that `docker pull docker.io/paketobuildpacks/builder-jammy-java-tiny:latest` will work but `curl -XPOST --unix-socket /var/run/docker.sock -H 'Content-Type: application/json' http://localhost/containers/create\?fromImage\=docker.io%2Fpaketobuildpacks%2Fbuilder-jammy-java-tiny%3Alatest` will fail with a `Enforced sign-in is enabled` response and a
```
missing X-Registry-Auth header
must be logged in to call /containers/create?fromImage=docker.io%2Fpaketobuildpacks%2Fbuilder-jammy-java-tiny%3Alatest
```
error in your `/Users/<USERNAME>/Library/Containers/com.docker.docker/Data/log/vm/01-docker.log` if you are on mac.

**Comment by philwebb** [comment]
For anyone else watching this issue, #45269 was merged in 3.5.0-RC1 which will hopefully solve some of the "enforced sign-in" problems.
